Add recommendations around variables usage (#52)
redhat-cop/automation-good-practices: inventories/README.adoc

+
TIP: if for some reason, you can't follow the recommendation, you can at least avoid duplicating too much information by indirectly referencing the hosts' variables as in `"{{ hostvars[item.name]['provision_value'] }}"`. Not so bad...
====

== Restrict your usage of variable types
[%collapsible]
====
Explanations::

> * Do not use playbook and play variables, as well as include_vars, prefer inventory variables.

Reviewer: ```suggestion
* Avoid playbook and play variables, as well as `include_vars`. 
  Opt for inventory variables instead.
```

Author: I won't comment on all your... comments but I'm taking them over in one big commit, and not one by one. I'm resolving them as I include them (without further... comment).

It also needs to be noted that many variable types tend to blur the separation between code and data and are to be avoided.
By reducing the number of variable types, you come to a more simple and overseeable list of variables, with explanations why they have their specific precedence, so that they become easier to remember and use wisely:
+
. role defaults (defined in `defaults/main.yml`), they are... defaults and can be overwritten by anything.
. inventory vars, they truly represent your desired state.
  They have their own internal precedence (group before host) but that's easy to remember.
. host facts don't represent a desired state but the current state, and no other variable should have the same name because of <<differentiate>> so that the precedence doesn't really matter.
. role vars (defined in `vars/main.yml`) represent constants used by the role to separate code from data, and shouldn't either collide with the inventory variables, but can be overwritten by extra vars if you know what you're doing.
. scoped vars, at the block or task level, are local to their scope and hence internal to the role, and can't collide with other variable types.

> . runtime vars, defined by register or set_facts, are taking precedence over almost everything defined previously, which makes sense as they represent the current state of the automation.

Reviewer: ```suggestion
. runtime vars, defined by `register` or `set_facts`, are taking precedence over almost everything defined previously, which makes sense as they represent the current state of the automation.
```

> . scoped params, at the role or include level this time, are admittedly a bit out of order and should be avoided to limit surprises.

Reviewer: ```suggestion
. scoped params, defined at the role or include level (as opposed to block or task level for scoped variables), are admittedly a bit out of order and should be avoided to limit surprises.
```

> . and extra_vars overwrite everything else (even runtime vars, which can be quite surprising)

Reviewer: ```suggestion
. and lastly, extra vars overwrite everything else (even runtime vars, which can be quite surprising)
```

> The following picture shows how this can be summarized in a simplified way to keep in mind, highlighting which variables are meant to overwrite each other:

Reviewer: ```suggestion
The following picture summarizes this list in a simplified and easier to keep in mind way, highlighting which variables are meant to overwrite others:
```

.Flow of variable precedences
image::variable_precedences.png[flow of variable precedences in 3 lanes]

CAUTION: even if we write that variables _shouldn't_ overwrite each other, they still all share the same namespace and _can_ potentially overwrite each other.
It is your responsibility as automation author to make sure they don't.
====

> == Prefer inventory variables for desired state variables over extra vars

Reviewer: ```suggestion
== Prefer inventory variables over extra vars to describe the desired state
```

[%collapsible]
====
Explanations::
Don't use extra vars to define your desired state.

> Make sure your inventory contains the complete state of your environment as you want to reach.

Reviewer: ```suggestion
Make sure your inventory completely describes how your environment is supposed to look like.
```

Use extra vars only for troubleshooting, debugging or validation purposes.

Rationale::
Inventory variables are typically in some kind of persistent tracked storage (be it a database or Git), and should be your sole source representing your desired state so that you can refer to it non-ambiguously.
On the other hand, extra vars are bound to a specific job or ansible-call and disappear together with history. 

> Examples::

Reviewer: Maybe add `no_log: "{{ overwrite_my_by_extra_var | default(true) }}"` as an example
